### PR TITLE
[FrameworkBundle] feature: add the ability to search a route

### PR DESCRIPTION
**Pull request by Simperfit**
| Q             | A
| ------------- | ---
| Branch?       | 4.1
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no 
| Tests pass?   | yes
| Fixed tickets | #26033 
| License       | MIT
| Doc PR        | symfony/symfony-docs#9236

This add the ability to search a route in the debug:router command.
img_3271

### CODE REVIEW

**Comment by Simperfit** [comment]
Travis failure is unrelated

**Comment by dunglas** [comment]
Can you add a test please?

**Comment by Simperfit** [comment]
@dunglas I will do, I'm searching a way to add one, I have to look on other command's tests to test the choices.

**Review by 20uf — APPROVED** [review]

**Comment by Simperfit** [comment]
Status: needs review

@dunglas test added

**Comment by Simperfit** [comment]
New test added that test the whole behaviour of the search.

File: src/Symfony/Bundle/FrameworkBundle/Command/RouterDebugCommand.php [hunk 1]
```diff
@@ -142,4 +150,17 @@ private function extractCallable(Route $route)
         } catch (\InvalidArgumentException $e) {
         }
     }
+
+    private function findRouteNameContaining(string $name, RouteCollection $routes)
```
Comment (dunglas) [inline src/Symfony/Bundle/FrameworkBundle/Command/RouterDebugCommand.php]: You can use the `: array` return type.

File: src/Symfony/Bundle/FrameworkBundle/Command/RouterDebugCommand.php [hunk 2]
```diff
@@ -79,6 +80,13 @@ protected function execute(InputInterface $input, OutputInterface $output)
         $routes = $this->router->getRouteCollection();
 
         if ($name) {
+            $matchingRoutes = $this->findRouteNameContaining($name, $routes);
+
+            if (!empty($matchingRoutes) && !$route = $routes->get($name)) {
```
Comment (dunglas) [inline src/Symfony/Bundle/FrameworkBundle/Command/RouterDebugCommand.php]: Can be simplified in=

```php
$matchingRoutes = $this->findRouteNameContaining($name, $routes) 
if ($matchingRoutes && !$route = $routes->get($name) {
```

File: src/Symfony/Bundle/FrameworkBundle/Command/RouterDebugCommand.php [hunk 3]
```diff
@@ -142,4 +150,17 @@ private function extractCallable(Route $route)
         } catch (\InvalidArgumentException $e) {
         }
     }
+
+    private function findRouteNameContaining(string $name, RouteCollection $routes)
+    {
+        $foundRoutesNames = array();
+        foreach ($routes as $routeName => $route) {
+            if (false === stripos($routeName, $name)) {
```
Comment (dunglas) [inline src/Symfony/Bundle/FrameworkBundle/Command/RouterDebugCommand.php]: Maybe can you invert the test, and put `$foundRoutesNames[] = $routeName;` in the if block to save 1 line (nitpicking).

**Review by dunglas — CHANGES_REQUESTED** [review]

File: src/Symfony/Bundle/FrameworkBundle/Command/RouterDebugCommand.php [hunk 4]
```diff
@@ -79,6 +80,12 @@ protected function execute(InputInterface $input, OutputInterface $output)
         $routes = $this->router->getRouteCollection();
 
         if ($name) {
+            $matchingRoutes = $this->findRouteNameContaining($name, $routes);
```
Comment (stof) [inline src/Symfony/Bundle/FrameworkBundle/Command/RouterDebugCommand.php]: I would first try to get the route by exact name before searching route names. It is useless to search the matching routes if we already have a valid name.

Comment (Simperfit) [reply src/Symfony/Bundle/FrameworkBundle/Command/RouterDebugCommand.php]: done

**Comment by Simperfit** [comment]
Status: Needs Review

**Comment by Simperfit** [comment]
PR rebased.

**Comment by chalasr** [comment]
@althaus that's a new feature, it must go on 4.1

**Comment by althaus** [comment]
@chalasr No plans for a 3.5? That's a pity for devs stuck on the 3.x branch for a while. 👎

**Comment by chalasr** [comment]
@alain-flaus Nope, 3.4 is the last 3.x version. That's the roadmap, nothing wrong, 4.0 is stable for a while.

**Comment by alain-flaus** [comment]
@chalasr Hello, I think I'm not the person you wanted to notify ?

**Comment by chalasr** [comment]
@alain-flaus sorry! edited

**Comment by chalasr** [comment]
I suggest to drop the existing `RouterDebugCommand` test, it is semi-functional (use a mocked kernel) and all of its tests are covered by the one added here.

**Comment by althaus** [comment]
@chalasr Oh... didn't know about the change in the release cycles and the limit of 5 minors.

So I'll just overload the command with my own version of this change.

**Comment by Simperfit** [comment]
@chalasr @dunglas PR rebased, test fixed, semi-functional test deleted.

**Comment by Simperfit** [comment]
Travis failure is unrelated.

File: src/Symfony/Bundle/FrameworkBundle/CHANGELOG.md [hunk 5]
```diff
@@ -17,7 +17,8 @@ CHANGELOG
    is either the service ID or the FQCN of the controller.
  * Deprecated `Symfony\Bundle\FrameworkBundle\Controller\ControllerNameParser`
  * The `container.service_locator` tag of `ServiceLocator`s is now autoconfigured.
-
+ * Add the ability to search a route in `debug:router`.
+
```
Comment (chalasr) [inline src/Symfony/Bundle/FrameworkBundle/CHANGELOG.md]: extra spaces I believe

**Review by chalasr — APPROVED** [review]

File: src/Symfony/Bundle/FrameworkBundle/Command/RouterDebugCommand.php [hunk 6]
```diff
@@ -76,7 +77,13 @@ protected function execute(InputInterface $input, OutputInterface $output)
         $routes = $this->router->getRouteCollection();
 
         if ($name) {
-            if (!$route = $routes->get($name)) {
+            if (!($route = $routes->get($name)) && $matchingRoutes = $this->findRouteNameContaining($name, $routes)) {
+                $default = 1 === count($matchingRoutes) ? $matchingRoutes[0] : null;
+                $name = $io->choice('Select one of the following routes to display its information', $matchingRoutes, $default);
```
Comment (fabpot) [inline src/Symfony/Bundle/FrameworkBundle/Command/RouterDebugCommand.php]: `Select one of the matching routes`?

**Review by fabpot — CHANGES_REQUESTED** [review]

**Comment by Simperfit** [comment]
@fabpot done.

File: src/Symfony/Bundle/FrameworkBundle/Command/RouterDebugCommand.php [hunk 7]
```diff
@@ -76,7 +77,13 @@ protected function execute(InputInterface $input, OutputInterface $output)
         $routes = $this->router->getRouteCollection();
 
         if ($name) {
-            if (!$route = $routes->get($name)) {
+            if (!($route = $routes->get($name)) && $matchingRoutes = $this->findRouteNameContaining($name, $routes)) {
+                $default = 1 === count($matchingRoutes) ? $matchingRoutes[0] : null;
+                $name = $io->choice('Select one of the matching routes to display its information', $matchingRoutes, $default);
```
Comment (fabpot) [inline src/Symfony/Bundle/FrameworkBundle/Command/RouterDebugCommand.php]: I think the `to display its information` is not needed.

Comment (Simperfit) [reply src/Symfony/Bundle/FrameworkBundle/Command/RouterDebugCommand.php]: I removed this part.

**Comment by Simperfit** [comment]
Status: Needs Review

**Review by fabpot — APPROVED** [review]

File: src/Symfony/Bundle/FrameworkBundle/Tests/Functional/RouterDebugCommandTest.php [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+        $ret = $tester->execute(array());
+        $display = $tester->getDisplay();
+
+        $this->assertSame(0, $ret, 'Returns 0 in case of success');
+        $this->assertContains('routerdebug_test', $display);
+        $this->assertContains('/test', $display);
+        $this->assertContains('/session', $display);
+    }
+
+    public function testDumpOneRoutes()
```
Comment (javiereguiluz) [inline src/Symfony/Bundle/FrameworkBundle/Tests/Functional/RouterDebugCommandTest.php]: Minor typo. `testDumpOneRoutes` -> `testDumpOneRoute`

**Review by javiereguiluz — APPROVED** [review]

File: src/Symfony/Bundle/FrameworkBundle/Tests/Functional/RouterDebugCommandTest.php (hunk 8 above)
Comment (Simperfit) [reply src/Symfony/Bundle/FrameworkBundle/Tests/Functional/RouterDebugCommandTest.php]: done

**Comment by fabpot** [comment]
Thank you @Simperfit.

**Pull request merged**